Tasks/DownloadBuildArtifacts/main.ts

```diff
@@ -0,0 +1,151 @@
+        if (definitionId) {
+            var build = await buildApi.getBuild(buildId, projectId).catch((reason) => {
+                reject(reason);
+                return;
+            });
+
+            if (build) {
+                if (build.definition.id !== parseInt(definitionId)) {
+                    reject(tl.loc("BuildIdBuildDefinitionMismatch", buildId, definitionId));
+                    return;
+                }
+            }
+            else {
+                reject(tl.loc("BuildNotFound", buildId));
+                return;
+            }
+        }
+
+        // populate itempattern and artifacts based on downloadType
+        if (downloadType === 'single') {
+            var artifactName = tl.getInput("artifactName");
+            var artifact = await buildApi.getArtifact(buildId, artifactName, projectId).catch((reason) => {
+                reject(reason);
+                return;
+            });
+
+            if (!artifact) {
+                reject(tl.loc("BuildArtifactNotFound", artifactName, buildId));
+                return;
+            }
+
+            artifacts.push(artifact);
+            itemPattern = artifactName + '/**';
+        }
+        else {
+            var buildArtifacts = await buildApi.getArtifacts(buildId, projectId).catch((reason) => {
+                reject(reason);
+            });
+
+            if (!buildArtifacts) {
+                tl.warning(tl.loc("NoArtifactsFound", buildId));
+                resolve();
+                return;
+            }
+
+            console.log(tl.loc("LinkedArtifactCount", buildArtifacts.length));
+            artifacts = artifacts.concat(buildArtifacts);
+            itemPattern = tl.getInput("itemPattern", false) || '**';
+        }
+
+        if (artifacts) {
+            var downloadPromises: Array<Promise<any>> = [];
+            artifacts.forEach(async function (artifact, index, artifacts) {
+                let downloaderOptions = new engine.ArtifactEngineOptions();
+                downloaderOptions.itemPattern = itemPattern;
+                downloaderOptions.verbose = isVerbose;
+
+                if (parallelLimit) {
+                    downloaderOptions.parallelProcessingLimit = parallelLimit;
+                }
+
+                if (artifact.resource.type.toLowerCase() === "container") {
+                    let downloader = new engine.ArtifactEngine();
+                    var containerParts: string[] = artifact.resource.data.split('/', 3);
+                    if (containerParts.length !== 3) {
+                        throw new Error(tl.loc("FileContainerInvalidArtifactData"));
+                    }
+
+                    var containerId: number = parseInt(containerParts[1]);
+                    var containerPath: string = containerParts[2];
+
+                    var itemsUrl = endpointUrl + "/_apis/resources/Containers/" + containerId + "?itemPath=" + containerPath + "&isShallow=true";
```

Comment:
Why are we specifying `&isShallow=true`?  Doesn't that mean we only get the immediate children at the path?

Reply:
O, nevermind, I see that we follow non-File items in processArtifactItemImplementation.